crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,512 @@
+        input_buffer.is_empty()
+    }
+
+    /// Read data from a `desc_chain` of the transmit queue and write it to the output sink.
+    ///
+    /// # Arguments
+    /// * `desc_chain` - A mutable reference to the descriptor chain that should point to the
+    ///                  buffers that the virtio console driver has sent to the device.
+    pub fn process_transmitq_chain<M>(
+        &mut self,
+        desc_chain: &mut DescriptorChain<M>,
+    ) -> Result<(), Error>
+    where
+        M: Deref,
+        M::Target: GuestMemory,
+    {
+        // Iterate through all the buffers in the chain and write the data to output.
+        while let Some(desc) = desc_chain.next() {
+            if desc.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor);
+            }
+            desc_chain
+                .memory()
+                .write_to(desc.addr(), &mut self.output, desc.len() as usize)
+                .map_err(Error::WriteToOutputFailed)?;
+
+            self.output.flush().map_err(Error::OutputSinkFlushFailed)?;
+        }
+
+        Ok(())
+    }
+
+    /// Takes data from the input buffer and writes it to guest memory in the specified
+    /// descriptor chain.
+    ///
+    /// The empty descriptor chain in which the data is written has previously
+    /// been added to the receive queue by the driver.
+    /// If there are not enough buffers to hold all the data, we stop and wait for the driver to
+    /// add more.
+    ///
+    /// # Arguments
+    /// * `desc_chain` - A mutable reference to the descriptor chain that should point to the
+    ///                  empty buffers that the virtio driver has offered the device.
+    /// # Returns
+    /// The number of raw bytes written to guest memory.
+    pub fn process_receiveq_chain<M>(
+        &self,
+        desc_chain: &mut DescriptorChain<M>,
+    ) -> Result<u32, Error>
+    where
+        M: Deref,
+        M::Target: GuestMemory,
+    {
+        // Acquire lock on buffer to make sure no other thread is adding data to it while it is
+        // being processed.
+        let mut input_buffer = self.input_buffer.lock().unwrap();
+        let mut sent_bytes: u32 = 0;
+
+        // When the driver adds a new descriptor chain to the receive queue this method will be
+        // called to write the remaining data to the newly provided buffers. If there is no
+        // remaining data waiting to be sent it can return 0.
+        if input_buffer.is_empty() {
+            return Ok(0);
+        }
+
+        // Break up the data from the "input_buffer" into chunks that fit into the buffers of the
+        // receive queue descriptor chain and write it to guest memory.
+        while let Some(desc) = desc_chain.next() {
+            if !desc.is_write_only() {
+                return Err(Error::UnexpectedReadOnlyDescriptor);
+            }
+            let len_write = cmp::min(desc.len() as usize, input_buffer.len());
```

Comment:
Since you are going to anyway push again to update the coverage value and to rebase, it would be clearer I think to convert input_buffer.len() here to u32, and have `len_write` as u32.